libtenzir/builtins/operators/ocsf.cpp

```diff
@@ -124,31 +131,62 @@ class caster {
   }
 
   auto cast_type(const type& ty) -> type {
+    if (ty.attribute("print_json")) {
+      if (print_json_) {
+        return type{string_type{}};
+      }
+      // We don't know the actual type, so we just use `null`.
+      return type{null_type{}};
+    }
     return match(ty, [&](const auto& ty) {
       return type{cast_type(ty)};
     });
   }
 
   template <basic_type Type>
-  auto cast(const type_to_arrow_array_t<Type>& array, const Type&,
-            std::string_view path) -> basic_series<Type> {
+  auto cast(basic_series<Type> input, const Type&, std::string_view path)
+    -> basic_series<Type> {
     (void)path;
-    return {Type{},
-            std::make_shared<type_to_arrow_array_t<Type>>(array.data())};
+    return input;
   }
 
-  auto cast(const arrow::Array& array, const type& ty, std::string_view path)
-    -> series {
-    if (is<string_type>(ty) and ty.attribute("print_json")) {
-      return series{
-        string_type{},
-        print_json(array, ty.attribute("nullify_empty_records").has_value())};
+  auto cast(series input, const type& ty, std::string_view path) -> series {
+    if (ty.attribute("print_json")) {
+      TENZIR_ASSERT(is<string_type>(ty));
+      if (ty.attribute("must_be_record")
+          and not(
+            input.type.kind().is_any<null_type, record_type>()
+            // Strings are also allowed so that `ocsf::apply` is idempotent.
+            or (print_json_ and input.type.kind().is<string_type>()))) {
+        diagnostic::warning("expected type `record` for `{}`, but got `{}`",
+                            path, input.type.kind())
+          .primary(self_)
+          .emit(dh_);
+        auto result_ty = print_json_ ? type{string_type{}} : type{null_type{}};
+        return series{result_ty, check(arrow::MakeArrayOfNull(
+                                   result_ty.to_arrow_type(), input.length()))};
+      }
+      if (print_json_) {
+        return series{
+          string_type{},
+          print_json(*input.array,
+                     ty.attribute("nullify_empty_records").has_value())};
+      }
+      // Otherwise, we just return the data exactly as we received it, without
+      // any further casting.
+      // TODO: Should we also nullify empty records here?
+      return input;
```

Comment:
We already set `unmapped` to `null` if it's `{}` when the option introduced by this PR is not set. Should this also happen with the option?

Reply:
I think this is generally useful.